featuremanagement/_featuremanager.py

```diff
@@ -70,14 +72,78 @@ def __init__(self, configuration, **kwargs):
         self._configuration = configuration
         self._cache = {}
         self._copy = configuration.get(FEATURE_MANAGEMENT_KEY)
-
         filters = [TimeWindowFilter(), TargetingFilter()] + kwargs.pop(PROVIDED_FEATURE_FILTERS, [])
 
         for feature_filter in filters:
             if not isinstance(feature_filter, FeatureFilter):
                 raise ValueError("Custom filter must be a subclass of FeatureFilter")
             self._filters[feature_filter.name] = feature_filter
 
+    @staticmethod
+    def _check_default_disabled_variant(feature_flag):
+        if not feature_flag.allocation:
+            return False
+        return FeatureManager._check_variant_override(
+            feature_flag.variants, feature_flag.allocation.default_when_disabled, False
+        )
+
+    @staticmethod
+    def _check_default_enabled_variant(feature_flag):
+        if not feature_flag.allocation:
+            return True
+        return FeatureManager._check_variant_override(
+            feature_flag.variants, feature_flag.allocation.default_when_enabled, True
+        )
+
+    @staticmethod
+    def _check_variant_override(variants, default_variant_name, status):
+        if not variants or not default_variant_name:
+            return status
+        for variant in variants:
+            if variant.name == default_variant_name:
+                if variant.status_override == "Enabled":
+                    return True
+                if variant.status_override == "Disabled":
+                    return False
+        return status
+
+    @staticmethod
+    def _is_targeted(context_id):
+        """Determine if the user is targeted for the given context"""
+        hashed_context_id = hashlib.sha256(context_id.encode()).hexdigest()
+        context_marker = abs(int(hashed_context_id, 16))
+        percentage = (context_marker / (2**256 - 1)) * 100
+
+        return percentage
+
+    def _assign_variant(self, feature_flag, **kwargs):
+        if not feature_flag.variants or not feature_flag.allocation:
+            return None
+        if feature_flag.allocation.user:
+            user = kwargs.get("user")
+            if user:
+                for user_allocation in feature_flag.allocation.user:
+                    if user in user_allocation.users:
+                        return user_allocation.variant
+        if feature_flag.allocation.group:
+            groups = kwargs.get("groups")
+            if groups:
+                for group_allocation in feature_flag.allocation.group:
+                    for group in groups:
+                        if group in group_allocation.groups:
+                            return group_allocation.variant
+        if feature_flag.allocation.percentile:
+            user = kwargs.get("user", "")
```

Comment:
Don't need a default because the default is None, right?

Reply:
The code below this will break if we don't have a String value. `""` is the assumed user if only a group is provided for allocation. Same if no user or group is provided?

Reply:
Which code below? A None will stop us from entering the if statement like all the other `get()` calls above.

Reply:
That None check isn't needed. That's why we have this issue.